### PR TITLE
Allow disable automated cleaning

### PR DESCRIPTION
**Pull request by fmuyassarov**
**What this PR does / why we need it**:
Implementation of [disabling disk cleaning
](https://github.com/metal3-io/metal3-docs/blob/master/design/cluster-api-provider-metal3/allow_disabling_node_disk_cleaning.md)

### CODE REVIEW

File: controllers/metal3machinetemplate_controller.go
Comment (namnx228) [inline controllers/metal3machinetemplate_controller.go:87]: Why do we need to have to reconcile function `reconcile` and `reconcileNormal`? Do you plan to have another reconcile function for another case?

**Review by namnx228 — COMMENTED** [review]
Overall, it looks good to me. Just one small question

File: controllers/metal3machinetemplate_controller.go
Comment (fmuyassarov) [reply controllers/metal3machinetemplate_controller.go:87]: Thanks @namnx228 . Currently we don't have `reconcileDelete()`, because we right don't do anything with M3MT with deletion timestamps. And, why `reconcileNormal` is because I just used the same reconciliation flow as in any other controller code. I can move all the logic under `reconcile` if you think it would be better, TBH I don't have strong opinion on that.

Comment (namnx228) [reply controllers/metal3machinetemplate_controller.go:87]: No, I also don't against the idea to have `reconcileNormal`. Let keep it like this.

Comment (fmuyassarov) [reply controllers/metal3machinetemplate_controller.go:87]: Okay, will do that. Thanks

File: controllers/metal3machinetemplate_controller.go
Comment (maelk) [inline controllers/metal3machinetemplate_controller.go:51]: you should handle the pause here

File: controllers/metal3machinetemplate_controller.go
Comment (maelk) [inline controllers/metal3machinetemplate_controller.go:111]: you need to call this in main.go

Comment (fmuyassarov) [reply controllers/metal3machinetemplate_controller.go:111]: Good catch. Fixed, thanks!

File: baremetal/metalmachinetemplate_manager.go
Comment (fmuyassarov) [reply baremetal/metalmachinetemplate_manager.go:61]: +1

File: baremetal/metalmachinetemplate_manager.go
Comment (fmuyassarov) [reply baremetal/metalmachinetemplate_manager.go:97]: good catch

File: api/v1alpha4/metal3machine_types.go
Comment (jan-est) [inline api/v1alpha4/metal3machine_types.go:67]: Maybe it would be good to add comment about what does default "metadata" value stands for? I understand that if `AutomatedCleaningMode` is set to "metadata" disk will be cleaned, but why not use disabled/enabled?

File: docs/api.md
Comment (furkatgofurov7) [inline docs/api.md:536]: We should also update [example M3MT](https://github.com/metal3-io/cluster-api-provider-metal3/pull/156/files#diff-9eddf4dc51bf6a0125ca7fb094468ad284112270385c41cebce4f8f0a29620abR559) with default value of `automatedCleaningMode` field.

File: api/v1alpha4/metal3machine_types.go
Comment (fmuyassarov) [reply api/v1alpha4/metal3machine_types.go:67]: Originally it was decided to be like this in BMO repo and the reason was in the future we might want to select particular disk from several to be deleted. For that reason, simply using disabled wasn't enough, because we can't specify which disk we would like to wipe out. Also, that was done to align with Ironic, because Ironic also has metadata option, and we might want to use that in the future.

File: docs/api.md
Comment (fmuyassarov) [reply docs/api.md:536]: Sorry, it seems URL is redirecting me to a commit, where shall I update it, in the document?

File: controllers/metal3machinetemplate_controller.go
Comment (fmuyassarov) [reply controllers/metal3machinetemplate_controller.go]: That part is removed completely

File: docs/api.md
Comment (furkatgofurov7) [reply docs/api.md:536]: Lines 548 (Example Metal3MachineTemplate) and onwards, bit down on this same document

File: baremetal/metalmachinetemplate_manager.go
Comment (fmuyassarov) [reply baremetal/metalmachinetemplate_manager.go]: Thanks, fixed

File: docs/api.md
Comment (fmuyassarov) [reply docs/api.md:537]: There is no spec.template part. Not sure if I should add a whole spec of Metal3Machine for showing cleaning field. Maybe only Metal3Machine example with `AutomatedCleaningMode` is enough?

File: api/v1alpha4/metal3machine_types.go
Comment (kashifest) [inline api/v1alpha4/metal3machine_types.go:68]: Why do we need this tag if we have `omitempty` two lines below?

File: baremetal/metalmachinetemplate_manager.go
Comment (kashifest) [inline baremetal/metalmachinetemplate_manager.go:57]: Is it only listing or also syncing the `AutomatedCleaningMode` ?

**Review by kashifest — APPROVED** [review]
Looks good to me. Couple of small nits. Please take care of the failed test.
/approve
/hold

File: api/v1alpha4/metal3machine_types.go
Comment (fmuyassarov) [reply api/v1alpha4/metal3machine_types.go:68]: My bad, forgot to clean that.

File: baremetal/metalmachinetemplate_manager.go
Comment (fmuyassarov) [reply baremetal/metalmachinetemplate_manager.go:57]: Good catch, fixing it

File: docs/api.md
Comment (furkatgofurov7) [reply docs/api.md:536]: I am not sure if I got your point. I was suggesting to put `AutomatedCleaningMode` field in example Metal3MachineTemplate as we did for nodeReuse field, i.e: 

```
      apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4 
      kind: Metal3MachineTemplate
      metadata:
      name: md-0
      spec:
        nodeReuse: false
        template:
          spec:
            AutomatedCleaningMode: metadata
            image:
              url: https://cloud-images.ubuntu.com/bionic/current/bionic-server-cloudimg-amd64.img
              checksum: https://cloud-images.ubuntu.com/bionic/current/bionic-server-cloudimg-amd64.img.md5sum
            hostSelector:
              matchLabels:
                key1: value1
              matchExpressions:
                key: key2
                operator: in
                values: {‘abc’, ‘123’, ‘value2’}
            dataTemplate:
              Name: md-0-metadata
```

Comment (fmuyassarov) [reply docs/api.md:536]: 🤦‍♂️, fixing it

File: baremetal/metalmachinetemplate_manager_test.go
Comment (furkatgofurov7) [inline baremetal/metalmachinetemplate_manager_test.go:212]: This seems to be copy of the previous [file](https://github.com/metal3-io/cluster-api-provider-metal3/blob/974a7fb98b0f5686d8cd9854b685c09f39de026d/baremetal/metal3machinetemplate_manager.go)?

File: controllers/metal3machinetemplate_controller.go
Comment (furkatgofurov7) [inline controllers/metal3machinetemplate_controller.go:35]: do we really need all of those verbs for status?

File: baremetal/metal3machinetemplate_manager_test.go
Comment (furkatgofurov7) [inline baremetal/metal3machinetemplate_manager_test.go:69]: nit: disabled

File: baremetal/metal3machinetemplate_manager_test.go
Comment (furkatgofurov7) [inline baremetal/metal3machinetemplate_manager_test.go:211]:
```suggestion
		Entry("Don't synchronize M3Machines which are not part of the M3MTemplate ", testCaseUpdate{
```

File: docs/api.md
Comment (furkatgofurov7) [inline docs/api.md:299]:
```suggestion
  `metadata`(default) value enables it. It is recommended to tune the cleaning via metal3MachineTemplate rather than
```

File: docs/api.md
Comment (furkatgofurov7) [inline docs/api.md:302]: For the sake of the consistency, maybe: 
```suggestion
  and eventually BaremetalHosts with the same value.
```

File: docs/api.md
Comment (furkatgofurov7) [inline docs/api.md:534]: Same `BaremetalHosts`

**Review by furkatgofurov7 — COMMENTED** [review]
Mostly lgtm, some suggestions for improvement:

File: baremetal/metalmachinetemplate_manager_test.go
Comment (fmuyassarov) [reply baremetal/metalmachinetemplate_manager_test.go:212]: Ops, two copies of the test code. Good catch

File: controllers/metal3machinetemplate_controller.go
Comment (fmuyassarov) [reply controllers/metal3machinetemplate_controller.go:35]: indeed not needed, we don't even have status subresource in the api. So, it doesn't make sense to keep this marker

**Comment by fmuyassarov** [comment]
@furkatgofurov7 Thanks for the reviews. I've addressed all the comments, PTAL.

**Comment by furkatgofurov7** [comment]
@fmuyassarov please squash your commits, I will lgtm after, thanks!

**Comment by fmuyassarov** [comment]
> @fmuyassarov please squash your commits, I will lgtm after, thanks!

I would like to keep commits separate to have clean history since each commit is doing different thing.

**Comment by furkatgofurov7** [comment]
> keep commits separate to have clean history

Ok, but I doubt that and IMO squashing the commits is the good way forward to maintain a clean git history.

**Review by furkatgofurov7 — COMMENTED** [review]
/lgtm

**Comment by fmuyassarov** [comment]
> > keep commits separate to have clean history
> 
> Ok, but I doubt that and IMO squashing the commits is the good way forward to maintain a clean git history.
I think it would be nice to have this documented somewhere, cause I can see some PRs getting squashed and some not. Having formal instructions would help us to be consistent.

**Pull request merged**
